db/wal_version.h

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+// This source code is licensed under both the GPLv2 (found in the
+// COPYING file in the root directory) and Apache 2.0 License
+// (found in the LICENSE.Apache file in the root directory).
+
+// WAL related classes used in VersionEdit and VersionSet.
+
+#pragma once
+
+#include <map>
+#include <ostream>
+#include <string>
+#include <vector>
+
+#include "logging/event_logger.h"
+#include "rocksdb/rocksdb_namespace.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+class JSONWriter;
+class Slice;
+class Status;
+
+using WalNumber = uint64_t;
+
+// Metadata of a WAL.
+class WalMetadata {
+ public:
+  WalMetadata() = default;
+
+  explicit WalMetadata(uint64_t size_bytes) : size_bytes_(size_bytes) {}
+
+  bool HasSize() const { return size_bytes_ != kUnknownWalSize; }
+
+  void SetSizeInBytes(uint64_t bytes) { size_bytes_ = bytes; }
+
+  uint64_t GetSizeInBytes() const { return size_bytes_; }
+
+ private:
+  // The size of WAL is unknown, used when the WAL is not closed yet.
+  constexpr static uint64_t kUnknownWalSize = 0;
+
+  // Size of a closed WAL in bytes.
+  uint64_t size_bytes_ = kUnknownWalSize;
+};
+
+// Records the event of adding a WAL in VersionEdit.
+class WalAddition {
+ public:
+  WalAddition() : number_(0), metadata_() {}
+
+  explicit WalAddition(WalNumber number) : number_(number), metadata_() {}
+
+  explicit WalAddition(WalNumber number, const WalMetadata& meta)
```

Comment:
You could also consider taking `meta` by value here and moving it into `metadata_`.